lib/src/modelinfo/modelinfo.dart

```diff
@@ -0,0 +1,167 @@
+    entities.forEach((e) {
+      if (e.model != model)
+        throw Exception("entity '${e.name}' with id ${e.id.toString()} has incorrect parent model reference");
+      if (lastEntityId.id < e.id.id)
+        throw Exception(
+            "lastEntityId ${lastEntityId.toString()} is lower than the one of entity '${e.name}' with id ${e.id.toString()}");
+      if (lastEntityId.id == e.id.id) {
+        if (lastEntityId.uid != e.id.uid)
+          throw Exception(
+              "lastEntityId ${lastEntityId.toString()} does not match entity '${e.name}' with id ${e.id.toString()}");
+        lastEntityIdFound = true;
+      }
+    });
+
+    if (entities.length > 0 && !lastEntityIdFound)
+      throw Exception("lastEntityId ${lastEntityId.toString()} does not match any entity");
+  }
+
+  Map<String, dynamic> toMap() {
+    Map<String, dynamic> ret = {};
+    ret["_note1"] = notes[0];
+    ret["_note2"] = notes[1];
+    ret["_note3"] = notes[2];
+    ret["entities"] = entities.map((p) => p.toMap()).toList();
+    ret["lastEntityId"] = lastEntityId.toString();
+    ret["lastIndexId"] = lastIndexId.toString();
+    ret["lastRelationId"] = lastRelationId.toString();
+    ret["lastSequenceId"] = lastSequenceId.toString();
+    ret["modelVersion"] = modelVersion;
+    ret["modelVersionParserMinimum"] = modelVersionParserMinimum;
+    ret["retiredEntityUids"] = retiredEntityUids;
+    ret["retiredIndexUids"] = retiredIndexUids;
+    ret["retiredPropertyUids"] = retiredPropertyUids;
+    ret["retiredRelationUids"] = retiredRelationUids;
+    ret["version"] = version;
+    return ret;
+  }
+
+  ModelEntity findEntityByUid(int uid) {
+    int idx = entities.indexWhere((e) => e.id.uid == uid);
+    return idx == -1 ? null : entities[idx];
+  }
+
+  ModelEntity findEntityByName(String name) {
+    final found = entities.where((e) => e.name.toLowerCase() == name.toLowerCase()).toList();
+    if (found.length == 0) return null;
+    if (found.length >= 2) throw Exception("ambiguous entity name: $name; please specify a UID in its annotation");
+    return found[0];
+  }
+
+  ModelEntity findSameEntity(ModelEntity other) {
+    ModelEntity ret;
+    if (other.id.uid != 0) ret = findEntityByUid(other.id.uid);
+    if (ret == null) ret = findEntityByName(other.name);
+    return ret;
+  }
+
+  ModelEntity createCopiedEntity(ModelEntity other) {
+    ModelEntity ret = createEntity(other.name, other.id.uid);
+    other.properties.forEach((p) => ret.createCopiedProperty(p));
+    return ret;
+  }
+
+  ModelEntity createEntity(String name, [int uid = 0]) {
+    int id = 1;
+    if (entities.length > 0) id = lastEntityId.id + 1;
+    if (uid != 0 && containsUid(uid)) throw Exception("uid already exists: $uid");
+    int uniqueUid = uid == 0 ? generateUid() : uid;
+
+    var entity = new ModelEntity(IdUid.create(id, uniqueUid), null, name, [], this);
+    entities.add(entity);
+    lastEntityId = entity.id;
+    return entity;
+  }
+
+  int generateUid() {
```

Comment:
Maybe add an explicit test for this? In case rng.nextInt ever changes, also the concat and bit shifting seems fragile (`<< 63` vs `<<64` commit :D).

Reply:
the generator will require further tests, that's for sure, I'll create an issue